### ISSUE TITLE
System.Text.Json of T Buggy JsonConverterOfT

### ISSUE THREAD

**Issue by vintzl** [dotnet/runtime#1153]
If I try to register many custom converters of `T`, in `ConfigureServices` or elsewhere, and I try to use the custom converter with a `CanConvert` overridden method, I get:

```Exception has occurred: CLR/System.InvalidOperationException
An unhandled exception of type 'System.InvalidOperationException' occurred in System.Text.Json.dll: 'The converter 'Buggy_JsonConverterOfT.CustomJsonConverter`1[System.Collections.Generic.List`1[Buggy_JsonConverterOfT.Customer]]' is not compatible with the type 'Buggy_JsonConverterOfT.Customer'.'
```
Indeed, even if the `CanConvert` method return `true`, the `JsonConverter` yield `false`…

Here a simple (drafty) repro:
- https://github.com/vintzl/Buggy_JsonConverterOfT

**Comment by ahsonkhan** [comment]
What are you trying to do here with the "buggy" converter and what do you expect the behavior to be? It is hard to parse what you are trying to do here. Does `JsonConverterFactory` work for your use  case?

https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverterfactory?view=netcore-3.1

**Comment by vintzl** [comment]
in `CanConvert` I want to put some logic, also to avoid [such issue](https://www.blackhat.com/docs/us-17/thursday/us-17-Munoz-Friday-The-13th-JSON-Attacks-wp.pdf).

I did try different registration order but I got the same result.

**Comment by layomia** [comment]
The `CanConvert` method of your generic custom converter [always returns `true`](https://github.com/vintzl/Buggy_JsonConverterOfT/blob/master/CustomJsonConverter.cs#L14) which means that you're indicating that the first concrete converter [you added to `JsonSerializerOptions.Converters`](https://github.com/vintzl/Buggy_JsonConverterOfT/blob/master/Startup.cs#L27) should be used for all types in the input object graph. This is what causes the exception you are seeing. Specifically, in your repro project, a [`CustomJsonConverter<List<Customer>>`](https://github.com/vintzl/Buggy_JsonConverterOfT/blob/master/Startup.cs#L27) converter which derives from `JsonConverter<List<Customer>>` cannot be used to [convert a `Customer` type](https://github.com/vintzl/Buggy_JsonConverterOfT/blob/master/Startup.cs#L30).

Please see the [docs](https://docs.microsoft.com/dotnet/standard/serialization/system-text-json-converters-how-to#steps-to-follow-the-factory-pattern) on how to follow the factory converter pattern which is ideal for writing generic converters.
